onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -1809,40 +1808,44 @@ class PlannerImpl {
+        // Neither trigger ActivateNotification/WaitOnEPStep for Shape op (whose output is ready for all the EPs), nor 
+        // upstream is on CPU device (As currently we never invoke RegisterWaitFn(CPU, ...) for all kinds of EP, thus no wait_handle can be retrieved for this case)
+        if (node->OpType() != "Shape" && execution_plan[i]->device_.Type() != OrtDevice::CPU) {
+          for (auto it = node->OutputNodesBegin(); it != node->OutputNodesEnd(); ++it) {
             for (auto* output : node->OutputDefs()) {
               if (output->Exists()) {
                 if (std::find(it->InputDefs().begin(), it->InputDefs().end(), output) != it->InputDefs().end()) {
                   OrtValueIndex output_arg_idx;
                   ORT_THROW_IF_ERROR(ort_value_name_idx_map_.GetIdx(output->Name(), output_arg_idx));
                   // there are two cases we need notification:
                   // 1. the consumer is not in the same stream
-                  // 2. the consumer is in the same stream(non-cpu device), but it consumer a CPU tensor from an non-shape op.
+                  // 2. the consumer is in the same stream(non-cpu device), but it consumes a CPU tensor from an non-shape op.
                   //    for example, a resize cuda kernel consumer a tensor from MemCpyToHost cuda kernel on the same stream.
                   //    in this case, the FIFO can't gurantee the cpu tensor is ready when resize kernel is launching
-                  if (node_stream_map_[it->Index()] != i ||
-                      (node_stream_map_[it->Index()] == i &&
-                       execution_plan[i]->device_.Type() != OrtDevice::CPU &&
-                       plan_.allocation_plan[output_arg_idx].location.device.Type() == OrtDevice::CPU)) {
-                    requires_notification = true;
-                    break;
+                  OrtDevice::DeviceType output_arg_device = plan_.allocation_plan[output_arg_idx].location.device.Type();
+                  WaitNotificationFn wait_handle = stream_handle_registry.GetWaitHandle(execution_plan[i]->device_.Type(), output_arg_device);
+                  if ((node_stream_map_[it->Index()] != i || output_arg_device == OrtDevice::CPU)
+                      && wait_handle != nullptr) {
+                    if (node_to_notification.find(node_index) == node_to_notification.end()) {
+                      node_to_notification[node_index] = plan_.notification_owners.size();
+                      plan_.notification_owners.push_back(i);
+                    }
+
+                    if (node_to_wait.find(it->Index()) == node_to_wait.end()) {
+                      node_to_wait[it->Index()] = std::vector<std::pair<WaitNotificationFn, NotificationIndex>>{std::make_pair(wait_handle, node_to_notification[node_index])};
+                    } else {
+                      node_to_wait[it->Index()].push_back(std::make_pair(wait_handle, node_to_notification[node_index]));
```

Comment:
if node B consume multiple outputs of node A, will you insert multiple wait handles to the same notification into this list?

Reply:
It will, as long as the wait_handle is valid under this case

Reply:
only 1 wait step is needed for this case